File: src/Cpp/18_Operacje_na_plikach/Zad10.cpp
#include <experimental/filesystem>
#include <iostream>
#include <string>

namespace filesys = std::experimental::filesystem;

//Otrzymujesz dwa napisy reprezentujace sciezki folderow.
//Skopiuj wszystkie pliki .png z pierwszego folderu do drugiego.

std::string nazwaPliku(const std::string& sciezka)
{
    return filesys::path(sciezka).filename();
}

void skopiujPlik(const std::string& sciezka, const std::string& sciezkaDocelowa)
{
    try {
        filesys::copy_file(sciezka, sciezkaDocelowa + filesys::path::preferred_separator + nazwaPliku(sciezka),
            filesys::copy_options::overwrite_existing);
    } catch (filesys::filesystem_error& e) {
        std::cout << e.what() << std::endl;
    }
}

std::string znajdzRozszerzenie(std::string sciezka)
{

    filesys::path obiekt(sciezka);

    if (obiekt.has_extension())
        return obiekt.extension().string();

    return "";
}

void skopiujPliki(const std::string& sciezka, const std::string& sciezkaDocelowa)
{

    for (const auto plik : filesys::directory_iterator(sciezka)) {
        if (znajdzRozszerzenie(plik.path()) == ".png")
            skopiujPlik(plik.path(), sciezkaDocelowa);
    }
}

int main()
{

    const std::string sciezka = "folder/";
    const std::string sciezkaDocelowa = "folder2/";
    skopiujPliki(sciezka, sciezkaDocelowa);

    return 0;
}

File: src/JavaScript/18_Operacje_na_plikach/Zad10.js
/*
Otrzymujesz dwa napisy reprezentujace sciezki folderow.
Skopiuj wszystkie pliki .png z pierwszego folderu do drugiego folderu.
*/

const fs = require('fs').promises;
const path = require('path');

skopiujPliki = async (zrodlo, cel, rozszerzenie) => {
    const dane = await fs.readdir(zrodlo);
    for (const danePliku of dane) {
        const sciezkaPliku = path.join(zrodlo, danePliku);
        const stat = await fs.lstat(sciezkaPliku);
        if (stat.isFile() && path.extname(danePliku) === rozszerzenie) {
            const sciezkaDocelowa = path.join(cel, danePliku);
            await fs.copyFile(sciezkaPliku, sciezkaDocelowa);
        }
    }
}

test1 = async () => {
    const sciezka1 = 'test1';
    const sciezka2 = 'test2';
    const sciezkiPlikow = ['plik1.png', 'plik2.png'];
    const tekst = 'przykladowy tekst.\n';

    await fs.mkdir(sciezka1);
    await fs.mkdir(sciezka2);

    for (const plik of sciezkiPlikow) {
        const sciezkaPliku = path.join(sciezka1, plik);
        await fs.writeFile(sciezkaPliku, tekst);
    }

    await skopiujPliki(sciezka1, sciezka2, '.png');

    for (const plik of sciezkiPlikow) {
        const sciezkaPliku = path.join(sciezka2, plik);
        oczekiwane = true;
        try {
            await fs.access(sciezkaPliku);
            wynik = true;
        } catch (err) {
            wynik = false;
        }
        if (wynik !== oczekiwane) {
            throw new Error(`Assertion error line 180: ${wynik} != ${oczekiwane}`);
        }
    }

    await fs.rmdir(sciezka1, {
        recursive: true
    });
    await fs.rmdir(sciezka2, {
        recursive: true
    });
}

main = async () => {
    await test1();
}

main();